src/Ionide.ProjInfo/Library.fs

```diff
@@ -291,120 +345,282 @@ module ProjectLoader =
+        | Error e -> Result.Error e
 
-            let compileItems = getCompileItems project
-            let nuGetRefs = getNuGetReferences project
-            let props = getProperties project properties
-            let sdkInfo = getSdkInfo props
-            let customProps = getProperties project customProperties
 
-            if not sdkInfo.RestoreSuccess then
-                Result.Error "not restored"
-            else
 
-                let proj = mapToProject path comandlineArgs p2pRefs compileItems nuGetRefs sdkInfo props customProps
 
-                Result.Ok proj
-        | Error e -> Result.Error e
+open Ionide.ProjInfo.Logging
+
+module WorkspaceLoaderViaProjectGraph =
+    let locker = obj ()
+
+
+type IWorkspaceLoader =
+    abstract member LoadProjects : string list * list<string> * bool -> seq<ProjectOptions>
+    abstract member LoadProjects : string list -> seq<ProjectOptions>
+    abstract member LoadSln : string -> seq<ProjectOptions>
+
+    [<CLIEvent>]
+    abstract Notifications : IEvent<WorkspaceProjectState>
+
+type WorkspaceLoaderViaProjectGraph private (toolsPath: ToolsPath) =
+    let logger = LogProvider.getLoggerFor<WorkspaceLoaderViaProjectGraph> ()
+    let loadingNotification = new Event<Types.WorkspaceProjectState>()
+
+
+
+    let handleProjectGraphFailures f =
+        try
+            f () |> Some
+        with :? Microsoft.Build.Exceptions.InvalidProjectFileException as e ->
+            let p = e.ProjectFile
+            loadingNotification.Trigger(WorkspaceProjectState.Failed(p, ProjectNotFound(p)))
+            None
+
+    let projectInstanceFactory projectPath globalProperties (projectCollection: ProjectCollection) =
+        let tfm = ProjectLoader.getTfm projectPath
+        ProjectInstance(projectPath, ProjectLoader.getGlobalProps projectPath tfm, null, projectCollection)
+
+    let projectGraphProjs (paths: string seq) =
+
+        handleProjectGraphFailures
+        <| fun () ->
+            paths |> Seq.iter (fun p -> loadingNotification.Trigger(WorkspaceProjectState.Loading p))
+            let entryPoints = paths |> Seq.map ProjectGraphEntryPoint
+            ProjectGraph(entryPoints, ProjectCollection.GlobalProjectCollection, projectInstanceFactory)
+
+    let projectGraphSln (path: string) =
+        handleProjectGraphFailures
+        <| fun () ->
+            let pg = ProjectGraph(path, ProjectCollection.GlobalProjectCollection, projectInstanceFactory)
+
+            pg.ProjectNodesTopologicallySorted
+            |> Seq.distinctBy (fun p -> p.ProjectInstance.FullPath)
+            |> Seq.map (fun p -> p.ProjectInstance.FullPath)
+            |> Seq.iter (fun p -> loadingNotification.Trigger(WorkspaceProjectState.Loading p))
+
+            pg
+
+
+
+
+
+
+    let loadProjects (projects: ProjectGraph, customProperties: string list, generateBinlog: bool) =
+        try
+            lock WorkspaceLoaderViaProjectGraph.locker
```

Comment:
Currently does a lock so we don't try to run more than one build (msbuild doesn't like that).  Could change this to a MailboxProcessor or Semaphor if desired.